test/integration/src/main/java/io/pravega/test/integration/utils/SecureSetupUtils.java

```diff
@@ -0,0 +1,238 @@
+    }
+
+    /**
+     * Start all pravega related services required for the test deployment.
+     *
+     * @throws Exception on any errors.
+     */
+    public void startAllServices() throws Exception {
+        startAllServices(null);
+    }
+
+    /**
+     * Start all pravega related services required for the test deployment.
+     *
+     * @param numThreads the number of threads for the internal client threadpool.
+     * @throws Exception on any errors.
+     */
+    public void startAllServices(Integer numThreads) throws Exception {
+        if (!this.started.compareAndSet(false, true)) {
+            log.warn("Services already started, not attempting to start again");
+            return;
+        }
+        this.clientConfig = generateValidClientConfig();
+        this.executor = ExecutorServiceHelpers.newScheduledThreadPool(2, "Controller pool");
+        this.controller = new ControllerImpl(ControllerImplConfig.builder().clientConfig(clientConfig).build(),
+                executor);
+        this.clientFactory = new ClientFactoryImpl(scope, controller, clientConfig);
+
+        // Start zookeeper.
+        this.zkTestServer = new TestingServerStarter().start();
+        this.zkTestServer.start();
+
+        // Start Pravega Service.
+        this.serviceBuilder = ServiceBuilder.newInMemoryBuilder(ServiceBuilderConfig.getDefaultConfig());
+
+        this.serviceBuilder.initialize();
+        StreamSegmentStore store = this.serviceBuilder.createStreamSegmentService();
+        this.server = new PravegaConnectionListener(false, servicePort, store, this.serviceBuilder.createTableStoreService(),
+                this.serviceBuilder.getLowPriorityExecutor());
+        this.server.startListening();
+        log.info("Started Pravega Service");
+
+        // Start Controller.
+        this.controllerWrapper = new ControllerWrapper(
+                this.zkTestServer.getConnectString(), false, true, controllerRPCPort, "localhost", servicePort,
+                Config.HOST_STORE_CONTAINER_COUNT, controllerRESTPort, this.authEnabled,
+                "../" + SecurityConfigDefaults.AUTH_HANDLER_INPUT_PATH, "secret", 600);
+        this.controllerWrapper.awaitRunning();
+        this.controllerWrapper.getController().createScope(scope).get();
+        log.info("Initialized Pravega Controller");
+    }
+
+    /**
+     * Stop the pravega cluster and release all resources.
+     *
+     * @throws Exception on any errors.
+     */
+    @Override
+    public void close() throws Exception {
+        if (!this.started.compareAndSet(true, false)) {
+            log.warn("Services not yet started or already stopped, not attempting to stop");
+            return;
+        }
+
+        this.controllerWrapper.close();
+        this.server.close();
+        this.zkTestServer.close();
+        this.serviceBuilder.close();
+        this.clientFactory.close();
+        this.controller.close();
+        this.executor.shutdown();
```

Comment:
This is the wrong way to shut down an executor. Use `ExecutorServiceHelpers.shutdown(...)`

Reply:
Using the method mentioned above.